In Windows host, use WriteConsoleW for stdout and stderr, and locale enabled print for file output

jtschuster opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/102174

Per https://github.com/dotnet/runtime/issues/102174#issuecomment-2109289546, this uses WriteConsoleW for writing to stdout and stderr, and the locale enabled _vfwprintf_l for writing to a file.

jtschuster committed:
Pass in locale with output methods

jtschuster committed:
Use WriteConsoleW for stdout and stderr so chcp isn't needed

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into DotnetInfoPath

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into DotnetInfoPath

jtschuster commented on src/native/corehost/hostmisc/pal.h:
```diff
@@ -159,13 +159,48 @@ namespace pal
     inline int str_vprintf(char_t* buffer, size_t count, const char_t* format, va_list vl) { return ::_vsnwprintf_s(buffer, count, _TRUNCATE, format, vl); }
     inline int strlen_vprintf(const char_t* format, va_list vl) { return ::_vscwprintf(format, vl); }
 
+    inline void file_vprintf(FILE* f, const char_t* format, va_list vl)
```
Is there a way to move them and keep them `inline`? I couldn't find a way to, so I've moved them and made them non-inline for now.

jtschuster commented on src/native/corehost/hostmisc/pal.h:
```diff
@@ -159,13 +159,48 @@ namespace pal
+
+    inline void out_vprintf(const char_t* format, va_list vl) {
+        // Get the length of the formatted string + newline + null terminator
+        int len = 2 + strlen_vprintf(format, vl);
```
Ah, I think so, that's probably what's causing the test failures.

jtschuster committed:
PR Feedback:

- Move print functions to pal.windows.cpp
- Rename print functions
- Use va_copy on the args when passed to two functions
- Add comment for why the methods are more complex

jtschuster commented:
Looks like the tests are failing because [WriteConsole fails](https://learn.microsoft.com/en-us/windows/console/writeconsole#:~:text=WriteConsole%20fails%20if%20it%20is%20used%20with%20a%20standard%20handle%20that%20is%20redirected%20to%20a%20file.) when stdout or stderr is redirected, so we need to write to a file in that case. Working on the fix.

jtschuster committed:
Write to stdout/err files when WriteConsole fails. Tests still failing

jtschuster committed:
Clean up print functions, use pal for print to file fallback

jtschuster commented:
There's a number of other string related functions in the pal that have locale variants, is there any reason not to update those to use UTF-8 everywhere in a follow-up?

elinor-fung reviewed (commented):
Is there a simple test we could add for this?
Maybe something similar to the below, but using `DotNetBuilder` to set up the install in a directory with characters that don't work properly without the change in this PR: 
https://github.com/dotnet/runtime/blob/8b1d06a5170628834b27c06b751ac0e54f65c7e3/src/installer/tests/HostActivation.Tests/DotnetArgValidation.cs#L87

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into DotnetInfoPath

jtschuster committed:
PR feedback

- move helper to anonymous namespace
- add test for utf-8/GB18030 characters
- add comments for why we need to use a locale

elinor-fung reviewed (approved):
Thanks!

jtschuster committed:
Formatting: move brace to new line
